The Slow March of Progress in Programming Language Tooling (#203)
earthly/website: blog/_posts/2022-07-11-programming-language-improvements.md

## Batteries-Included Standard Library

> I'm not sure whether it's a blessing or a curse, but expansive standard libraries [are a big improvement]. It's possible to get a lot done with PHP, Python, and Go without needing to install any third-party library. They come with json, http client and server, and even database access for the most part.
>
> Amir Saeid

### Standard Library History

A standard library is a library of common things that comes with the language. C has `libc`, and C++ has `libcpp`, but both feel very minimal compared to the common conception of a batteries-included standard library.

The history is a little unclear to me, but it seems like Python (1991) was the first language to take the stance of really having an extensive standard library. Java 1.0 (1996) also came with an extensive standard library (The Java Class Library), and many other languages would follow suit.

It's hard to overstate the benefits of getting simple things done without implementing foundational stuff yourself or having to reach for third-party dependencies.

### Standard Library State of the Art: GoLang

> Most modern languages (that aren't JavaScript) now ship with extensive standard libraries. In particular, though, Go puts special emphasis on its standard library. It promises backward compatibility, performance, and well-thought-out implementations. Because of this, go developers lean on the standard library more than many other communities and generally hold it in high regard.

Reviewer: ```suggestion
Most modern languages (that aren't JavaScript) now ship with extensive standard libraries. In particular, though, Go puts special emphasis on its standard library. It promises backward compatibility, performance, and well-thought-out implementations. Because of this, Go developers lean on the standard library more than many other communities and generally hold it in high regard.
```

<div class="notice notice--big">
## Side Note: Backporting.

{% picture content-wide-nocrop {{site.pimages}}{{page.slug}}/8470.png --alt {{ Backporting }} %}

Ok, so you might wonder, if CPAN made Perl better and every language moving forward embraced package managers for third-party code, why didn't languages predating this add on package managers after the fact?

Well, they did, but it doesn't seem obvious how to get agreement in a language community after a certain point. I'm not sure why this is, but maybe people don't like change?

This is going to be a running theme. Once the idioms and patterns of a programming language, and a language community, have been established its hard to go back and add things on. At least I think this is why Javascript has NPM, Rust has crates but C++ has [dds, cpm, conan, pacm, spakc, buckaroo, hunter and vcpkg](https://stackoverflow.com/questions/27866965/does-c-have-a-package-manager-like-npm-pip-gem-etc). You know what is better than one universally agreed upon package manager? Not eight of them!

> One counter point on backporting here though is standard libraries. C++ added its standard library rather late in the game, by successfully bring parts of STL under its roof. So existing languages can adopting tooling innovations, its just a bigger lift.

Reviewer: ```suggestion
One counter point on backporting here though is standard libraries. C++ added its standard library rather late in the game, by successfully bringing parts of STL under its roof. So existing languages can adopt tooling innovations, its just a bigger lift.
```

</div>

So after CPAN, a great standard library that does most of the things you need to do, augmented with a third-party package repository that is simple to use and straight forward to contribute to became the table stakes for a language.

## Documentation Support

{% picture content-wide-nocrop {{site.pimages}}{{page.slug}}/6680.png --alt {{ Javadoc }} %}

> ## Documentation: History

Reviewer: ```suggestion
### Documentation: History
```

> Once you have third-party packages, you need an easy way to document them. Javadoc and it's generated documentation was the first version of this I encountered. It made it much easier to find what I was looking in the Java Class Library because I could just click around in the Javadocs on the web. Then if you combine Javadoc with IDE integration and it becomes simple to work with code you've never seen before. Exploratory coding becomes possible.

Reviewer: ```suggestion
Once you have third-party packages, you need an easy way to document them. Javadoc and it's generated documentation was the first version of this I encountered. It made it much easier to find what I was looking for in the Java Class Library because I could just click around in the Javadocs on the web. Then you can combine Javadoc with IDE integration and it becomes simple to work with code you've never seen before. Exploratory coding becomes possible.
```

> ## Documentation: State of the Art

Reviewer: ```suggestion
### Documentation: State of the Art
```

## Write Once, Run Most Places ™

> [One improvement I've seen has been] J2EE and the standardization of web servers have unlocked the computing foundation on which our civilization depends today.
> Java and the JVM pioneered that but I don't think they are being adequately credited for it. Once Java became a thing, the platform you develop on stopped being relevant with regards to the platform you deploy on. It's obvious knowledge today, but 20 years ago, this was revolutionary.
>
> Cédric Beust

### Run Anywhere: History

{% picture content-wide-nocrop {{site.pimages}}{{page.slug}}/7310.png --alt {{ Java ME on Nokia Phones }} %}

Java and the JVM really pushed the line for cross platform development. No longer did your developer environment need to match your production environment. With the JVM, you could compile things down to JARs and run it anywhere that had a Java Virtual Machine installed.

Virtualization and then containerization have made this far more common, but the Java was the first major programming language to enable this type of run anywhere workflow.

### Run Anywhere: State of the Art

> There are of course downsides to the Java approach. One of which is the slow start-up time of JIT code and another is the limitation of not being able to easily call something that isn't written in Java. GraalVM claims these problems have be overcome but the current trend seems to be ahead-of-time cross-compiling. Both Rust and Go make this fairly simple, assuming you don't have any c or `libc` dependendcies.

Reviewer: ```suggestion
There are of course downsides to the Java approach. One of which is the slow start-up time of JIT code and another is the limitation of not being able to easily call something that isn't written in Java. GraalVM claims these problems have been overcome but the current trend seems to be ahead-of-time cross-compiling. Both Rust and Go make this fairly simple, assuming you don't have any c or `libc` dependencies.
```

### Package Manager: State of the Art

>I think one main advantage that cargo has is that it came with the language. Retro active build tools often lack integration into the platform as a whole
>
> Robert Masen

This package manager area is fast-moving. If you invest a lot of engineering hours, you can improve your language onboarding and day to day usage experience substantially. And so the stakes here keep growing. The `cargo` and `rustup` documentation for Rust are almost as extensive as the rust book, and that doesn't include all the cargo plugins.

Can you easily switch compiler versions in your language? Can you quickly run tests? Can you easily do code coverage, and performance testing? Vendor code? Generate documentation? Lint code? Fix the code lints? All of these used to be stand-alone tools in various language ecosystems and now they come with Rust out of the box. And it's a similar story for Go. I assume other newer languages will try to match or exceed this level of tooling.

## Code Formatters

## Case Study: `gofmt`

> Code formatters existed before `gofmt`, just as third-party software packages exited before `CPAN`, but making something a standard for a community profoundly changes things. For example, no language that was created before Go is likely to achieve the near 100% style conformance that Go has because the existing languages have to deal with existing code, whereas `gofmt` enforces a single style and has no knobs to tweak. Languages that follow Go can learn this lesson. And so Rust (`rustfmt`) and Zig (`zig fmt`) have embraced a strong default code style and an accompanying code formatter, and have gained an edge in developer experience because of this.

Reviewer: ```suggestion
Code formatters existed before `gofmt`, just as third-party software packages existed before `CPAN`, but making something a standard for a community profoundly changes things. For example, no language that was created before Go is likely to achieve the near 100% style conformance that Go has because the existing languages have to deal with existing code, whereas `gofmt` enforces a single style and has no knobs to tweak. Languages that follow Go can learn this lesson. And so Rust (`rustfmt`) and Zig (`zig fmt`) have embraced a strong default code style and an accompanying code formatter, and have gained an edge in developer experience because of this.
```

(Tell me what I missed and/or got wrong, and I'll do an update.)

<div class="notice--info">
### The Things That Didn't Spread

There are also the things that language creators were sure would change the world, but yet never gained wider adoption, or only became a fundamental expectation in one particular niche.

Certainly, Jupyter notebooks and REPLs are essential in some domains but unknown in others. Even more niche are Smalltalk's image-based approach to state and Mathematica / Wolfram language's language-integrated-data.
</div>

## Conclusion

The tooling that enables developers to get work done is a massive part of what makes a language useable. Tooling is also a changing landscape where standards keep rising.

It goes like this:

When a new developer tooling innovation is discovered, newer programming languages get a chance to bake that innovation into their language tooling. Doing so gives them an incremental advantage, and these increments add up over time to a better developer experience.

> So newer languages have one clear well thought out way to do something and older languages will have either many contradictory ways, or no ways at all, do the same thing. And this makes older languages feel old.

Reviewer: ```suggestion
So newer languages have one clear, well thought out way to do something, and older languages will have either many contradictory ways, or no ways at all, to do the same thing. And this makes older languages feel old.
```
